### ISSUE TITLE
Add warning when KL-divergence is negative

### ISSUE THREAD

**Issue by lvwerra** [lvwerra/trl#235]
A common failure mode of the training is that the generation kwargs are not correctly set which leads to negative KL-divergence which is almost always a precursor for failed training. We should at least show a warning when this happens.

**Comment by janpawlowskiof** [comment]
Would you care to explain how is the negative KL-divergence caused by incorrect generation kwargs? 

I am trying to train a fine-tuned polish T5 model with kwargs `{"top_k": 0.0, "top_p": 1.0, "do_sample": True, "eos_token_id": 1, "max_new_tokens": 128}`, but objective/kl immediately becomes negative and then everything goes to hell, and i cannot figure out what is the thing I'm missing.
![image](https://user-images.githubusercontent.com/17483828/226899673-b0a95765-ad92-4a01-b891-7b70ced9dc38.png)

**Comment by younesbelkada** [comment]
From my understanding and discussions with @lvwerra , manually adding `eos_token_id:-1` is needed in order for `generate` to continue generation. Otherwise whenever the model predicts `1` as the next token, the generation will not continue, thus leads to much shorter sequences sometimes and being the rootcause of unusual training behaviors

**Comment by janpawlowskiof** [comment]
Thanks for your response! 
Unfortunately I did try this as I was slightly desperate, but with no effect. I got exactly the same behavior. Also I have a preview of generated sentences and they are rather long, so I do not think that this is the cause of my issues.

**Comment by lvwerra** [comment]
Does it work if you just plug in your model into the T5 example? Trying to figure out the source of the issue (model or code). Or vice versa plugin the model used in the T5 example into your code. Even if the rewards don't make much sense with the switched language the negative KL shouldn't occur in any case.

**Comment by janpawlowskiof** [comment]
![image](https://user-images.githubusercontent.com/17483828/227198311-c6663f4a-3522-4a27-83fe-4356f4badc6d.png)

Blue plot is my dataset, purple plot is original dataset from t5-sentiment sample (like it says in the legend).

From what it seems, this is at least somewhat dependent on the dataset that I am using. I am unsure if I can share the dataset, but I will try to figure this out and one I have a while to spare, I will maybe try to create a minimal example where this happens.

**Comment by lvwerra** [comment]
It's a bit hard to know what's going on exactly. In general, a negative KL means that the reference model assigns a higher probability to the sampled tokens than the policy which was used to sample it, which means something with sampling during generation is off.

Usually I look at log-probs of the generated tokens and compare active to reference models to see where the issue might be.

**Comment by janpawlowskiof** [comment]
This might be a silly question, but why calculate kl as
`kl = logprob - ref_logprob` and not `kl = (logprob - ref_logprob).abs()`?

Wouldn't it make more sense to give penalty based on the absolute value of this difference?

**Comment by ycsun1972** [comment]
I also had the same problem. In my case, I turned on eval mode for the actor model when doing sampling. I fixed this problem and everything worked fine.
@janpawlowskiof

**Comment by younesbelkada** [comment]
This sounds great! @janpawlowskiof can you double check?

>  In my case, I turned on eval mode for the actor model when doing sampling. I fixed this problem and everything worked fine.

@edbeeching this sounds like the solution you suggested us offline to disable dropout

**Comment by Alymostafa** [comment]
@ycsun1972 Can you elaborate on how to do that in the code?

**Comment by janpawlowskiof** [comment]
Unfortunately, this does not seem to fix my issues. 
Whether model is in train/eval does influence its logprobs, which is definitely noteworthy, but I still have problems with stability (kl not being able to reach target value, going negative or exploding), even when using a constant reward o 1

**Comment by GauravVirmani** [comment]
@janpawlowskiof For me, reducing the learning rate worked. 

I noticed that a higher value of learning rate leads to very high variance in KL, sometimes causing it to become negative or exploding.

**Comment by lvwerra** [comment]
@janpawlowskiof what kind of model are you using and what are you generation settings?

**Comment by vwxyzjn** [comment]
Hello all, 
I have a [60-lines-of-code snippet](https://gist.github.com/vwxyzjn/6f967bbadfec2ff3b19ac0b2c1ba5503) that can reproduce the negative KL divergence issue.

4 out of 9 experiments experienced negative KL and subsequently crashed, and the issue does not always appear.

<img width="1154" alt="Screenshot 2023-06-07 at 8 02 20 AM" src="https://github.com/lvwerra/trl/assets/5555347/3b77b0cc-cc43-40d4-ba71-c079852c9406">

<img width="895" alt="Screenshot 2023-06-07 at 8 05 40 AM" src="https://github.com/lvwerra/trl/assets/5555347/fc129f5b-6329-4b58-ae1c-c0c5f947de8e">

**Comment by vwxyzjn** [comment]
Doing further investigation. It turns out the `min_length`, `eos_token_id` and `max_new_tokens` made a difference


```diff
 generation_kwargs = {
-    "min_length": len(desired_tensor[0]),
+    "min_length": -1,
     "top_k": 0.0,
     "top_p": 1.0,
     "do_sample": True,
     "pad_token_id": tokenizer.eos_token_id,
+    "eos_token_id": -1,
+    "max_new_tokens": len(desired_tensor[0]),
 }
```

<img width="1187" alt="image" src="https://github.com/lvwerra/trl/assets/5555347/b1041c29-b8d3-4ec6-ae45-2e4c93afebc1">

**Comment by Ignoramus0817** [comment]
@vwxyzjn Thansk for your work!
I am also suffering from negative KL divergence problem using the default parameters in the script. According to your investigation, if I want to make sure the training is stable, I should set min_length=1, eos_token_id=-1 . Am I right?
 
Besides, when I tried this setting, I got a warning``eos_token_id` has to be a list of positive integers, but is [-1]`. Should I just ignore that? And in inference time, should I also set eos_token_id to be -1? Thanks in advance.

**Comment by vwxyzjn** [comment]
@Ignoramus0817 unfortunately this negative KL divergence seems like a much more common issue. See https://github.com/lvwerra/trl/issues/417. I am still doing some investigation :)

**Comment by lvwerra** [comment]
#487 should fix the issues in #417.

**Comment by lvwerra** [comment]
Closing this as the original issue was resolved and the neg-KL problem is covered in other issues.

**Comment by lih1130** [comment]
Hi, I have a question about calculating KLD. I saw `logprobs_from_logits()` function, there is something reason to use labels rather than model logit probability? I understand that to obtain KLD between ref_model and model, we need to use softmax of logit instead of label.

**Comment by Anonymousplendid** [comment]
@vwxyzjn Thank you for your great repro and experiments. However, can you elaborate on why these three parameters make a difference and how to correctly set them since I use a generation kwargs of {"min_length":-1, "eos_token_id":-1, "max_new_token_id": 128(also tried 512)} and failed?

**Comment by lvwerra** [comment]
@lih1130 the reason we use the label is because we only compute the KL w.r.t. the generated token (not all possible tokens). We use the label to select the logprob of the generated token.

@Anonymousplendid You can read about it in the docs https://huggingface.co/docs/trl/how_to_train#what-is-the-concern-with-negative-kl-divergence

**Comment by Anonymousplendid** [comment]
> @lih1130 the reason we use the label is because we only compute the KL w.r.t. the generated token (not all possible tokens). We use the label to select the logprob of the generated token.
> 
> @Anonymousplendid You can read about it in the docs https://huggingface.co/docs/trl/how_to_train#what-is-the-concern-with-negative-kl-divergence

Thank you for your response. However, the post does not answer why max_new_tokens affect kl problem. Besides, with the hyperparameters above changed, my experiment still failed (offering gibberish) even though plots are good.

**Comment by seanie12** [comment]
I don't know why but if i set gradient_accumulation_steps > 1, I got the same error. Otherwise it is fine.
